File: core/src/main/resources/sudoku_generator.py
import random
import mysql.connector
from copy import deepcopy

# ============================================================

# CONFIG

# ============================================================

MYSQL_HOST = "localhost"
MYSQL_USER = "root"
MYSQL_PASSWORD = ""
MYSQL_DATABASE = "sudoku"

NUM_6X6 = 100
NUM_9X9 = 100

# ============================================================

# MYSQL

# ============================================================

conn = mysql.connector.connect(
host=MYSQL_HOST,
user=MYSQL_USER,
password=MYSQL_PASSWORD,
database=MYSQL_DATABASE
)

cursor = conn.cursor()

# ============================================================

# GENERIC SUDOKU HELPERS

# ============================================================

def get_box_size(size):
if size == 9:
return (3, 3)
elif size == 6:
return (2, 3)
else:
raise Exception("Unsupported size")

def is_valid(board, row, col, num, size):
if num in board[row]:
return False

```
for r in range(size):
    if board[r][col] == num:
        return False

box_rows, box_cols = get_box_size(size)

start_row = row - row % box_rows
start_col = col - col % box_cols

for r in range(start_row, start_row + box_rows):
    for c in range(start_col, start_col + box_cols):
        if board[r][c] == num:
            return False

return True
```

def find_empty(board, size):
for r in range(size):
for c in range(size):
if board[r][c] == 0:
return (r, c)
return None

# ============================================================

# SOLVER

# ============================================================

def solve(board, size):
empty = find_empty(board, size)

```
if not empty:
    return True

row, col = empty

nums = list(range(1, size + 1))
random.shuffle(nums)

for num in nums:
    if is_valid(board, row, col, num, size):
        board[row][col] = num

        if solve(board, size):
            return True

        board[row][col] = 0

return False
```

# ============================================================

# COUNT SOLUTIONS

# ============================================================

def count_solutions(board, size, limit=2):
empty = find_empty(board, size)

```
if not empty:
    return 1

row, col = empty

count = 0

for num in range(1, size + 1):
    if is_valid(board, row, col, num, size):
        board[row][col] = num

        count += count_solutions(board, size, limit)

        board[row][col] = 0

        if count >= limit:
            return count

return count
```

# ============================================================

# GENERATE FULL SOLUTION

# ============================================================

def generate_complete_board(size):
board = [[0] * size for _ in range(size)]
solve(board, size)
return board

# ============================================================

# MAKE UNIQUE PUZZLE

# ============================================================

def create_puzzle(solution_board, size):
puzzle = deepcopy(solution_board)

```
cells = [(r, c) for r in range(size) for c in range(size)]
random.shuffle(cells)

for row, col in cells:
    backup = puzzle[row][col]
    puzzle[row][col] = 0

    temp = deepcopy(puzzle)

    solutions = count_solutions(temp, size)

    if solutions != 1:
        puzzle[row][col] = backup

return puzzle
```

# ============================================================

# SERIALIZE

# ============================================================

def board_to_string(board):
return ''.join(
str(cell)
for row in board
for cell in row
)

# ============================================================

# INSERT

# ============================================================

def insert_puzzle(size, puzzle, solution):
cursor.execute(
"""
INSERT INTO sudoku_puzzles
(
size,
grid,
solution
)
VALUES
(
%s,
%s,
%s
)
""",
(
size,
puzzle,
solution
)
)

# ============================================================

# GENERATION LOOP

# ============================================================

def generate_and_insert(size, count):
print(f"Generating {count} puzzles for size {size}")

```
for i in range(count):
    solution_board = generate_complete_board(size)

    puzzle_board = create_puzzle(
        solution_board,
        size
    )

    puzzle_str = board_to_string(
        puzzle_board
    )

    solution_str = board_to_string(
        solution_board
    )

    insert_puzzle(
        size,
        puzzle_str,
        solution_str
    )

    conn.commit()

    print(
        f"Inserted {i + 1}/{count} size {size}"
    )
```

# ============================================================

# MAIN

# ============================================================

generate_and_insert(6, NUM_6X6)
generate_and_insert(9, NUM_9X9)

print("Done")

cursor.close()
conn.close()

File: generator/src/main/java/com/funfriday/game/generator/sudoku/SudokuGenerator.java
package com.funfriday.game.generator.sudoku;

import com.funfriday.game.generator.Generator;
import com.funfriday.game.generator.GeneratorSchedule;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

@Slf4j
@Component("sudokuGenerator")
@GeneratorSchedule(interval = "PT12H")
public class SudokuGenerator implements Generator {

    private static final String URL = "jdbc:mysql://localhost:3306/GameData";
    private static final String USER = "root";
    private static final String PASSWORD = "";

    @Override
    public void generate() throws Exception {
        try (Connection conn = DriverManager.getConnection(URL, USER, PASSWORD)) {
            log.info("Starting sudoku generation cycle");
            generateAndInsert(conn, 6);
            generateAndInsert(conn, 9);
            log.info("Completed sudoku generation cycle");
        } catch (Exception e) {
            log.error("Sudoku generation cycle failed", e);
            throw e;
        }
    }

    private static void generateAndInsert(Connection conn, int size) throws Exception {
        String sql = "INSERT INTO sudoku_puzzles(size, grid, solution) VALUES (?, ?, ?)";
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            try {
                log.info("Generating sudoku puzzle size: " + size);
                int[][] solution = generateSolvedBoard(size);
                int[][] puzzle = createPuzzle(solution, size);

                ps.setInt(1, size);
                ps.setString(2, boardToString(puzzle));
                ps.setString(3, boardToString(solution));
                ps.executeUpdate();
                log.info("Inserted sudoku puzzle size: " + size);
            } catch (Exception e) {
                log.error("Failed to insert sudoku puzzle size " + size, e);
                throw e;
            }
        }
    }

    private static int[][] generateSolvedBoard(int size) {
        int[][] board = new int[size][size];
        solve(board, size, new Random());
        return board;
    }

    private static boolean solve(int[][] board, int size, Random rnd) {
        int[] empty = findEmpty(board, size);
        if (empty == null) return true;

        List<Integer> nums = new ArrayList<>();
        for (int i = 1; i <= size; i++) nums.add(i);
        Collections.shuffle(nums, rnd);

        int r = empty[0], c = empty[1];
        for (int n : nums) {
            if (isValid(board, r, c, n, size)) {
                board[r][c] = n;
                if (solve(board, size, rnd)) return true;
                board[r][c] = 0;
            }
        }
        return false;
    }

    private static int[][] createPuzzle(int[][] solution, int size) {
        int[][] puzzle = copy(solution);
        List<int[]> cells = new ArrayList<>();
        for (int r = 0; r < size; r++) {
            for (int c = 0; c < size; c++) {
                cells.add(new int[]{r, c});
            }
        }
        Collections.shuffle(cells);

        for (int[] cell : cells) {
            int r = cell[0], c = cell[1];
            int backup = puzzle[r][c];
            puzzle[r][c] = 0;

            int[][] tmp = copy(puzzle);
            if (countSolutions(tmp, size, 2) != 1) {
                puzzle[r][c] = backup;
            }
        }
        return puzzle;
    }

    private static int countSolutions(int[][] board, int size, int limit) {
        int[] empty = findEmpty(board, size);
        if (empty == null) return 1;

        int r = empty[0], c = empty[1];
        int count = 0;

        for (int n = 1; n <= size; n++) {
            if (isValid(board, r, c, n, size)) {
                board[r][c] = n;
                count += countSolutions(board, size, limit);
                board[r][c] = 0;

                if (count >= limit) return count;
            }
        }
        return count;
    }

    private static boolean isValid(int[][] board, int row, int col, int num, int size) {
        for (int i = 0; i < size; i++) {
            if (board[row][i] == num || board[i][col] == num) return false;
        }

        int boxRows = size == 9 ? 3 : 2;
        int boxCols = size == 9 ? 3 : 3;

        int sr = row - row % boxRows;
        int sc = col - col % boxCols;

        for (int r = sr; r < sr + boxRows; r++) {
            for (int c = sc; c < sc + boxCols; c++) {
                if (board[r][c] == num) return false;
            }
        }
        return true;
    }

    private static int[] findEmpty(int[][] board, int size) {
        for (int r = 0; r < size; r++) {
            for (int c = 0; c < size; c++) {
                if (board[r][c] == 0) return new int[]{r, c};
            }
        }
        return null;
    }

    private static int[][] copy(int[][] src) {
        int[][] out = new int[src.length][src[0].length];
        for (int i = 0; i < src.length; i++) {
            System.arraycopy(src[i], 0, out[i], 0, src[i].length);
        }
        return out;
    }

    private static String boardToString(int[][] board) {
        StringBuilder sb = new StringBuilder();
        for (int[] row : board) {
            for (int v : row) sb.append(v);
        }
        return sb.toString();
    }
}
